src/Streamly/Internal/FileSystem/Event.hs

```diff
@@ -0,0 +1,236 @@
+#elif defined(CABAL_OS_WINDOWS)
+import Streamly.Internal.FileSystem.Event.Windows (Event)
+import qualified Streamly.Internal.FileSystem.Event.Windows as Event
+#else
+#error "FS Events not supported on this platform"
+#endif
+
+-- XXX Ensure that defaultConfig is the same for all platforms.
+--
+-- XXX Ensure that equivalent events are generated on all platforms when same
+-- operations are done. Document any differences in behavior on different
+-- platforms. Any events that are not common to all platforms should be
+-- filtered out.
+--
+-- XXX Only need to set common event types for macOS/Linux/Windows
+-- XXX write a test, that generates all types of events that can possibly
+-- occur on each platform. Then create a common test that only checks
+-- for the common subset of those.
+
+-------------------------------------------------------------------------------
+-- Creating a watch
+-------------------------------------------------------------------------------
+
+-- macOS and Windows seem to do a path level watch i.e. they watch the path
+-- rather than the underlying inode, possibly they can map the inode back to a
+-- path?.  On Linux we cannot keep watching a watch root if the inode it is
+-- pointing to has moved. However, the inode continues to be watched in this
+-- case instead of the path, which is not the case for macOS.
+--
+-- To make the behavior same for all platforms, for the Linux case we will have
+-- to stop watching the inode if it moves from under the path
+-- (setUnwatchMoved).  And on macOS we can stop watching  a path if the inode
+-- under it changes i.e. just filter out events for that path.
+--
+-- Also, to make the behavior of Linux same as macOS we need to set
+-- setFollowSymLinks on Linux.
+
+-- XXX Verify all the cases mentioned below using test cases.
+
+-- | Start monitoring a list of file system paths for file system events.  The
+-- paths could be files or directories.  Monitoring starts from the current
+-- time onwards. The paths are specified as UTF-8 encoded 'Array' of 'Word8'.
+--
+-- If a watch root is a symbolic link then the target of the link is watched.
+-- Fails if the watched path does not exist. If the user does not have
+-- permissions (specific permissions?) on the watch root then no events are
+-- generated.  No events are generated if the watch root itself is renamed or
+-- deleted.
+--
+-- Note: not yet implemented on macOS, use watchRecursive instead.
+--
+-- /Pre-release/
+--
+watch :: NonEmpty (Array Word8) -> SerialT IO Event
+#if defined(CABAL_OS_WINDOWS)
+watch = Event.watch
+#elif defined(CABAL_OS_LINUX)
+watch = Event.watchWithFlags
+    [ 0x00000002    -- File was modified.
+    , 0x00000004    -- Metadata changed.
+    , 0x00000100    -- Subfile was created.
+    , 0x00000200    -- Subfile was deleted.
+    , 0x00000400    -- Root was deleted.
+    , 0x00000040    -- File was moved from X.
+    , 0x00000080    -- File was moved to Y.
+    ]
```

Comment:
This is not right, it makes the defaultConfig of windows module different than what `watch` is using, which is terrible. We should set the flags in the `defaultConfig` of the windows module.